File: tun_asym_cksum.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <fcntl.h>
#include <sys/ioctl.h>
#include <linux/if.h>
#include <linux/if_tun.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>

// Constants for TUN interface.
#define IFF_TUN 0x0001
#define IFF_NO_PI 0x1000

int create_tun_interface(const char* name) {
    struct ifreq ifr;
    int fd = open("/dev/net/tun", O_RDWR);

    if (fd < 0) {
        perror("Opening /dev/net/tun");
        exit(1);
    }

    memset(&ifr, 0, sizeof(ifr));
    ifr.ifr_flags = IFF_TUN | IFF_NO_PI;
    strncpy(ifr.ifr_name, name, IFNAMSIZ);

    if (ioctl(fd, TUNSETIFF, (void*)&ifr) < 0) {
        perror("ioctl(TUNSETIFF)");
        close(fd);
        exit(1);
    }

    return fd;
}

void configure_interface(const char* name, const char* address) {
    char command[256];
    snprintf(command, sizeof(command), "ip addr add %s dev %s", address, name);
    system(command);
    snprintf(command, sizeof(command), "ip link set dev %s up", name);
    system(command);
}

unsigned short checksum(void* b, int len) {
    unsigned short* buf = b;
    unsigned int sum = 0;
    unsigned short result;

    for (sum = 0; len > 1; len -= 2)
        sum += *buf++;
    if (len == 1)
        sum += *(unsigned char*)buf;
    sum = (sum >> 16) + (sum & 0xFFFF);
    sum += (sum >> 16);
    result = ~sum;

    return result;
}

/*void send_packet(char* ip_packet, int packet_len, char* dest_ip) {
    int sock = socket(AF_INET, SOCK_RAW, IPPROTO_RAW);
    if (sock < 0) {
        perror("Socket creation failed");
        exit(1);
    }

    struct sockaddr_in dest;
    memset(&dest, 0, sizeof(dest));
    dest.sin_family = AF_INET;
    dest.sin_addr.s_addr = inet_addr(dest_ip);

    if (sendto(sock, ip_packet, packet_len, 0, (struct sockaddr*)&dest, sizeof(dest)) < 0) {
        perror("Sendto failed");
    }

    close(sock);
}

void respond_to_ping(int tun_fd) {
    printf("Listening for ICMP packets...\n");
    char packet[2048];

    while (1) {
        int nread = read(tun_fd, packet, sizeof(packet));
        if (nread < 0) {
            perror("Reading from tun interface");
            break;
        }

        if (packet[20] == 8) {  // ICMP Echo Request
            printf("Received an ICMP Echo Request\n");

            // Create ICMP Echo Reply
            packet[20] = 0;  // Echo Reply
            unsigned short* icmp_header = (unsigned short*)(packet + 20);
            icmp_header[1] = 0; // Reset checksum to 0
            icmp_header[1] = checksum(icmp_header, nread - 20);

            // Swap source and destination IP
            for (int i = 0; i < 4; i++) {
                char temp = packet[12 + i];
                packet[12 + i] = packet[16 + i];
                packet[16 + i] = temp;
            }

            // Extract destination IP
            char dest_ip[INET_ADDRSTRLEN];
            inet_ntop(AF_INET, packet + 16, dest_ip, INET_ADDRSTRLEN);

            // Send the response packet
            printf("Sending response packet...\n");
            send_packet(packet, nread, dest_ip);  // Use nread as packet length
        }
    }
}*/

void send_packet(int sock, char* ip_packet, int packet_len, char* dest_ip) {
    struct sockaddr_in dest;
    memset(&dest, 0, sizeof(dest));
    dest.sin_family = AF_INET;
    dest.sin_addr.s_addr = inet_addr(dest_ip);

    if (sendto(sock, ip_packet, packet_len, 0, (struct sockaddr*)&dest, sizeof(dest)) < 0) {
        perror("Sendto failed");
    }
}

void respond_to_ping(int tun_fd) {
    printf("Listening for ICMP packets...\n");
    char packet[2048];

    // Open the socket once
    int sock = socket(AF_INET, SOCK_RAW, IPPROTO_RAW);
    if (sock < 0) {
        perror("Socket creation failed");
        exit(1);
    }

    while (1) {
        int nread = read(tun_fd, packet, sizeof(packet));
        if (nread < 0) {
            perror("Reading from tun interface");
            break;
        }

        if (packet[20] == 8) {  // ICMP Echo Request
            printf("Received an ICMP Echo Request\n");

            // Create ICMP Echo Reply
            packet[20] = 0;  // Echo Reply
            unsigned short* icmp_header = (unsigned short*)(packet + 20);
            icmp_header[1] = 0; // Reset checksum to 0
            icmp_header[1] = checksum(icmp_header, nread - 20);

            // Swap source and destination IP
            for (int i = 0; i < 4; i++) {
                char temp = packet[12 + i];
                packet[12 + i] = packet[16 + i];
                packet[16 + i] = temp;
            }

            // Extract destination IP
            char dest_ip[INET_ADDRSTRLEN];
            inet_ntop(AF_INET, packet + 16, dest_ip, INET_ADDRSTRLEN);

            // Send the response packet
            printf("Sending response packet...\n");
            send_packet(sock, packet, nread, dest_ip);  // Use nread as packet length
        }
    }

    // Close the socket when done
    close(sock);
}

int main() {
    int tun_fd = create_tun_interface("tunX");
    configure_interface("tunX", "10.0.0.10/24");

    printf("tunX interface created and configured. Listening for ICMP packets...\n");

    respond_to_ping(tun_fd);

    close(tun_fd);
    return 0;
}

File: tun_asym_cksum.py
import os
import fcntl
import struct
import socket
from pyroute2 import IPRoute

# Constants for TUN interface
TUNSETIFF = 0x400454ca
IFF_TUN = 0x0001
IFF_NO_PI = 0x1000

def create_tun_interface(name='tunX'):
    # Open the TUN device file using os.open
    tun_fd = os.open('/dev/net/tun', os.O_RDWR)
    
    # Create the TUN interface
    ifr = struct.pack('16sH', name.encode('utf-8'), IFF_TUN | IFF_NO_PI)
    fcntl.ioctl(tun_fd, TUNSETIFF, ifr)
    
    return tun_fd

def configure_interface(name='tunX', address='10.0.0.10/24'):
    ip = IPRoute()
    idx = ip.link_lookup(ifname=name)[0]
    ip.addr('add', index=idx, address=address.split('/')[0], mask=int(address.split('/')[1]))
    ip.link('set', index=idx, state='up')

def send_packet(ip_packet, dest_ip):
    # Create a raw socket with AF_INET
    sock = socket.socket(socket.AF_INET, socket.SOCK_RAW, socket.IPPROTO_RAW)
    try:
        # Send the packet
        sock.sendto(ip_packet, (dest_ip, 0))
    finally:
        sock.close()

def checksum(source_string):
    """
    Calculate the checksum of the input string
    """
    sum = 0
    max_count = (len(source_string) // 2) * 2
    count = 0
    while count < max_count:
        val = source_string[count + 1] * 256 + source_string[count]
        sum = sum + val
        sum = sum & 0xffffffff  # Necessary?
        count = count + 2

    if max_count < len(source_string):
        sum = sum + source_string[len(source_string) - 1]
        sum = sum & 0xffffffff  # Necessary?

    sum = (sum >> 16) + (sum & 0xffff)
    sum = sum + (sum >> 16)
    answer = ~sum
    answer = answer & 0xffff
    answer = answer >> 8 | (answer << 8 & 0xff00)
    return answer

def respond_to_ping(tun_fd):
    print("Listening for ICMP packets...")
    while True:
        # Read a packet from the TUN interface
        packet = os.read(tun_fd, 2048)
        # Check if it's an ICMP echo request (ping)
        if packet[20] == 8:  # ICMP type 8 is Echo Request
            print("Received an ICMP Echo Request")
            # Create an ICMP Echo Reply
            icmp_type = 0  # Echo Reply
            icmp_code = 0
            icmp_checksum = 0
            icmp_id = packet[24:26]
            icmp_seq = packet[26:28]
            icmp_payload = packet[28:]
            icmp_header = struct.pack('!BBHHH', icmp_type, icmp_code, icmp_checksum, *struct.unpack('!HH', icmp_id + icmp_seq))
            icmp_checksum = checksum(icmp_header + icmp_payload)
            icmp_header = struct.pack('!BBHHH', icmp_type, icmp_code, icmp_checksum, *struct.unpack('!HH', icmp_id + icmp_seq))

            # Swap source and destination IP addresses
            ip_header = packet[:20]
            swapped_ip_header = ip_header[:12] + ip_header[16:20] + ip_header[12:16]

            # Create final packet
            response = swapped_ip_header + icmp_header + icmp_payload

            # Extract destination IP
            dest_ip = socket.inet_ntoa(packet[12:16])

            # Send the response packet via appropriate interface
            print("Sending response packet via ens33")
            send_packet(response, dest_ip)

def main():
    tun_fd = create_tun_interface()
    configure_interface()
    try:
        print("tunX interface created and configured. Listening for ICMP packets...")
        respond_to_ping(tun_fd)
    except KeyboardInterrupt:
        print("Stopping...")
    finally:
        os.close(tun_fd)

if __name__ == '__main__':
    main()
